src/main/scala/com/fasterxml/jackson/module/scala/ser/EitherSerializer.scala

```diff
@@ -1,95 +1,174 @@
+      case Some(vts) => ser.serializeWithType(content, jgen, provider, vts)
+      case None => ser.serialize(content, jgen, provider)
     }
+    jgen.writeEndObject()
+  }
 
-    val typeSer = valueTypeSerializer.map(_.forProperty(property))
-    val candidateSerializer = tryContentSerializer(prov, property, serializerFromAnnotation(property, prov).orElse(elementSerializer))
+  override def serializeWithType(value: Either[AnyRef, AnyRef], jgen: JsonGenerator, provider: SerializerProvider, typeSer: TypeSerializer): Unit = {
+    if (value == null) {
+      provider.defaultSerializeNull(jgen)
+      return
+    }
 
-    if (candidateSerializer != elementSerializer || property != beanProperty.orNull || valueTypeSerializer != typeSer)
-      new EitherSerializer(elementType, typeSer, Option(property), candidateSerializer.asInstanceOf[Option[JsonSerializer[AnyRef]]])
-    else this
+    // Otherwise apply type-prefix/suffix, then std serialize:
+    typeSer.writeTypePrefixForObject(value, jgen, classOf[Either[AnyRef, AnyRef]])
+    serialize(value, jgen, provider, Some(typeSer))
+    typeSer.writeTypeSuffixForObject(value, jgen)
   }
 
+  protected[this] def findCachedSerializer(prov: SerializerProvider, typ: Class[_]): JsonSerializer[AnyRef] = {
+    var ser = dynamicSerializers.serializerFor(typ)
+    if (ser == null) {
+      ser = findSerializer(prov, typ, property)
+      dynamicSerializers = dynamicSerializers.newWith(typ, ser)
+    }
+    ser
+  }
 }
 
-
 private object EitherSerializerResolver extends Serializers.Base {
 
   private val EITHER = classOf[Either[AnyRef, AnyRef]]
-
-  override def findSerializer(config: SerializationConfig, javaType: JavaType, beanDesc: BeanDescription) =
-    if (!EITHER.isAssignableFrom(javaType.getRawClass)) null else {
-      val elementType = Option(javaType.containedType(0))
-      val typeSer = elementType.flatMap(e => Option(e.getTypeHandler).map(_.asInstanceOf[TypeSerializer]))
-      val valSer = elementType.flatMap(e => Option(e.getValueHandler).map(_.asInstanceOf[JsonSerializer[AnyRef]]))
-
-      new EitherSerializer(elementType, typeSer, None, valSer)
+  private val LEFT = classOf[Left[AnyRef, AnyRef]]
+  private val RIGHT = classOf[Right[AnyRef, AnyRef]]
+
+  override def findReferenceSerializer(config: SerializationConfig,
+                                       refType: ReferenceType,
+                                       beanDesc: BeanDescription,
+                                       contentTypeSerializer: TypeSerializer,
+                                       contentValueSerializer: JsonSerializer[AnyRef]): JsonSerializer[_] = {
+    if (!EITHER.isAssignableFrom(refType.getRawClass)) null
+    else {
+      val javaType = if (LEFT.isAssignableFrom(refType.getRawClass) || RIGHT.isAssignableFrom(refType.getRawClass)) {
+        refType.getReferencedType.getSuperClass
+      } else refType.getReferencedType
+
+      val leftType = javaType.containedType(0)
```

Comment:
Yes, although this is new with 2.7. Earlier it was done as modification after construction, but there were some reasons why making it "float up" makes more sense... I forget the exact details now, but it seemed either impossible or impractical to make it work after the fact, at the end.

On plus side making them match from inside simplifies code quite a bit as only exact equality check is needed (so 'if (clz == Either.class) { ... }` and not if instanceof).